File: src/SIR.Client.Web/LiveApi.fs
namespace SIR.Client.Web

open Fable.Core
open Fable.Core.JsInterop
open SIR.Protocol.Http

[<RequireQualifiedAccess>]
module LiveApi =

    [<Global>]
    let private fetch (url: string, options: obj) : JS.Promise<obj> = jsNative

    let bootstrap (request: BootstrapV1.Request) =
        async {
            let options =
                createObj
                    [ "method" ==> "POST"
                      "headers" ==>
                        createObj
                            [ "Content-Type" ==> "application/json"
                              "X-SIR-Development-Actor" ==> request.ActorName ]
                      "body" ==> BootstrapV1.encodeRequest request ]

            let! response = fetch ("/api/bootstrap", options) |> Async.AwaitPromise
            let! body = response?text() |> Async.AwaitPromise

            if not (unbox<bool> response?ok) then
                failwith $"bootstrap request failed: {body}"

            return
                BootstrapV1.responseFromJson (string body)
                |> Result.defaultWith (fun error -> failwith $"bootstrap response did not decode: {error}")
        }

File: src/SIR.Client.Web/.fable-rules/LiveApi.js

import { singleton } from "./fable_modules/fable-library-js.5.13.0/AsyncBuilder.js";
import { responseFromJson, encodeRequest } from "./SIR.Protocol/Http.js";
import { awaitPromise } from "./fable_modules/fable-library-js.5.13.0/Async.js";
import { Exception } from "./fable_modules/fable-library-js.5.13.0/Util.js";
import { Result_DefaultWith } from "./fable_modules/fable-library-js.5.13.0/Result.js";
import { concat } from "./fable_modules/fable-library-js.5.13.0/String.js";
import { toString } from "./fable_modules/fable-library-js.5.13.0/Types.js";

export function bootstrap(request) {
    return singleton.Delay(() => {
        const options = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "X-SIR-Development-Actor": request.ActorName,
            },
            body: encodeRequest(request),
        };
        return singleton.Bind(awaitPromise(fetch("/api/bootstrap", options)), (_arg) => {
            const response = _arg;
            return singleton.Bind(awaitPromise(response.text()), (_arg_1) => {
                const body = _arg_1;
                return singleton.Combine(!response.ok ? (((() => {
                    throw new Exception(`bootstrap request failed: ${body}`);
                })(), singleton.Zero())) : singleton.Zero(), singleton.Delay(() => singleton.Return(Result_DefaultWith((error) => {
                    throw new Exception(concat("bootstrap response did not decode: ", error));
                }, responseFromJson(toString(body))))));
            });
        });
    });
}

